docs/design.md

```diff
@@ -0,0 +1,371 @@
+## CARv2 + DAG store implementation notes
+
+## Overview
+
+The purpose of the CARv2 + DAG store endeavour is to eliminate overhead from the
+deal-making processes with the mission of unlocking scalability, performance,
+and resource frugality on both miner and client side within the Filecoin
+network.
+
+Despite serving Lotus/Filecoin immediate needs, we envision the DAG store to be
+a common interplanetary building block across IPFS, Filecoin, and IPLD-based
+projects.
+
+The scalability of the former two is bottlenecked by the usage of Badger as a
+monolithic blockstore. The DAG store represents an alternative that leverages
+sharding and the concept of detachable "data cartridges" to feed and manage data
+as transactional units.
+
+For the project that immediately concerns us, the definition of done is the
+removal of the Badger blockstore from all client and miner codepaths involved in
+storage and retrieval deals, and the transition towards interfacing directly
+with low-level data repository files known as CARs (Content ARchives), on both
+the read and the write sides.
+
+Note that Badger is generally a fitting database for write-intense and
+iterator-heavy workloads. It just doesn't withstand the pattern of usage we
+subject it to in large IPLD-based software components, especially past the 100s
+GBs data volume mark.
+
+For more information on motivation and architecture, please refer to
+[CAR-native DAG store](https://docs.google.com/document/d/118fJdf8HGK8tHYOCQfMVURW9gTF738D0weX-hbG-BvY/edit#). This document is 
+recommended as a pre-read.
+
+## Overview
+
+The DAG store comprises three layers:
+
+1. Storage layer (manages shards)
+2. Index repository (manages indices)
+3. DAG access layer (manages queries)
+
+## Storage layer
+
+The DAG store holds shards of data. In Filecoin, shard = deal. Each shard is a
+repository of data capable of acting as a standalone blockstore.
+
+### Shards
+
+Shards are identified by an opaque byte string: the shard key. In the case of
+Filecoin, the shard key is the `PieceCID` of the storage deal.
+
+A shard contains:
+
+1. the shard key (identifier).
+2. the means to obtain shard data (a `Mount` that provides the CAR either
+   locally, remotely, or some other location).
+3. the shard index (both acting as a manifest of the contents of the shard, and
+   a means to efficiently perform random reads).
+
+### CAR formats
+
+A shard can be filled with CARv1 and CARv2 data. CARv2 can be indexed or
+indexless.
+
+The choice of version and characteristics affects how the shard index is
+populated:
+
+1. CARv1: the index is calculated upon shard activation.
+2. Indexless CARv2: the index is calculated upon shard activation.
+3. Indexed CARv2: the inline index is adopted as-is as the shard index.
+
+### Shard states
+
+1. **Available:** the system knows about this shard and is capable of serving
```

Comment:
The shard state is never exposed to the DAG store client, right ? This state is only used for internal bookkeeping by the DAG Store. 

So, for example, when the user calls the DAG access layer for retrieval by a shard key, the DAG store will look up the shard state to determine if it's Available. If it's not available, the dag access layer can ask the scrap area to fetch the  CAR file (by using the `Mount` to Stat/Info/Fetch etc.) and then mark the shard state as `Available`.

Basically, there will be no `IsAvailable` or `MarkAvailable` etc APIs as the details of the shard state and how that state is changed are hidden by the DAG Store and only used for internal bookkeeping.

Reply:
Exactly 👍  We should provide status accessors to query the state of a shard from the outside, but only on an informational level.